rasa/core/policies/rule_policy.py

```diff
@@ -294,69 +268,69 @@ def train(
         trackers_as_actions = rule_trackers_as_actions + story_trackers_as_actions
 
         # negative rules are not anti-rules, they are auxiliary to actual rules
-        form_unhappy_lookup = self._create_form_unhappy_lookup_from_states(
+        self.lookup[
+            RULES_FOR_FORM_UNHAPPY_PATH
+        ] = self._create_form_unhappy_lookup_from_states(
             trackers_as_states, trackers_as_actions
         )
-        self.lookup[RULES_FOR_FORM_UNHAPPY_PATH] = self._clean_feature_keys(
-            form_unhappy_lookup
-        )
 
         # TODO use story_trackers and rule_trackers
         #  to check that stories don't contradict rules
 
         logger.debug(f"Memorized '{len(self.lookup[RULES])}' unique rules.")
 
     @staticmethod
-    def _features_in_state(features: List[Text], state: Dict[Text, float]) -> bool:
-
-        state_slots = defaultdict(set)
-        for s in state.keys():
-            if s.startswith("slot"):
-                state_slots[s[: s.rfind("_")]].add(s)
-
-        f_slots = defaultdict(set)
-        for f in features:
-            # TODO: this is a hack to make a rule know
-            #  that slot or form should not be set;
-            #  `_None` is added inside domain to indicate that
-            #  the feature should not be present
-            if f.endswith("_None"):
-                if any(f[: f.rfind("_")] in key for key in state.keys()):
-                    return False
-            elif f not in state:
-                return False
-            elif f.startswith("slot"):
-                f_slots[f[: f.rfind("_")]].add(f)
+    def _check_rule_state(rule_state: STATE, state: STATE) -> bool:
+
+        for state_type, rule_sub_state in rule_state.items():
+            sub_state = state.get(state_type, {})
```

Comment:
optional: Makes it a bit easier to distinguish between rule and other state .
```suggestion
            conversation_sub_state = state.get(state_type, {})
```